ethereum/sidecar/server.go

```diff
@@ -428,6 +537,342 @@ func (s *Server) startGRPCServer(
+		if seqStart.LT(sdkmath.OneInt()) {
+			seqStart = sdkmath.OneInt()
+		}
+
+		events, err := s.bridgeOutClient.GetAssetsUnlockedEvents(
+			ctx,
+			seqStart,
+			seqEnd,
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to fetch AssetsUnlocked events [%s, %s): %w",
+				seqStart.String(),
+				seqEnd.String(),
+				err,
+			)
+		}
+
+		// We know how many events we should receive. Any difference suggests
+		// an error on `mezod` server side (e.g. gaps in stored event database).
+		expectedLength := seqEnd.Sub(seqStart).Int64()
+		if int64(len(events)) != expectedLength {
+			return nil, fmt.Errorf(
+				"fetched unexpected number of AssetsUnlocked events for "+
+					"range [%s, %s); expected %d, got %d",
+				seqStart.String(),
+				seqEnd.String(),
+				expectedLength,
+				len(events),
+			)
+		}
+
+		for i := len(events) - 1; i >= 0; i-- {
+			event := events[i]
+			if event.BlockTime < cutOffBlockTime {
+				// We found an event older than cut-off time.
+				break outer
+			}
+			recentEvents = append(recentEvents, event)
+		}
+
+		// If the length of fetched events was smaller than the batch size,
+		// there are no more events to fetch.
+		if len(events) < s.assetsUnlockedBatchSize {
+			break
+		}
+
+		sequenceTip = seqStart.SubRaw(1)
+	}
+
+	// Ensure events are sorted in ascending order of their unlock sequences.
+	sort.Slice(recentEvents, func(i, j int) bool {
+		return recentEvents[i].UnlockSequence.LT(recentEvents[j].UnlockSequence)
+	})
+
+	return recentEvents, nil
+}
+
+// findUnattestedAssetsUnlockedEvents finds unattested `AssetsUnlocked` events
+// from the passed input events. An event is considered attested if there
+// was an `AssetsUnlockConfirmed` event emitted for it on Ethereum and its
+// unlock sequence is among the confirmed unlocks.
+func (s *Server) findUnattestedAssetsUnlockedEvents(
+	ctx context.Context,
+	inputEvents []bridgetypes.AssetsUnlockedEvent,
+) ([]bridgetypes.AssetsUnlockedEvent, error) {
+	// If there are no input events, return early with an empty list.
+	if len(inputEvents) == 0 {
+		return []bridgetypes.AssetsUnlockedEvent{}, nil
+	}
+
+	// Use unlock sequences of the input events as a filter.
+	unlockSequences := make([]*big.Int, len(inputEvents))
+	for i, event := range inputEvents {
+		unlockSequences[i] = event.UnlockSequence.BigInt()
+	}
+
+	// Finalized block is considered safe from reorgs as it is 64-96 blocks
+	// behind the current tip. It can be used as a filter.
+	finalizedBlock, err := s.chain.FinalizedBlock(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the finalized block: [%w]", err)
+	}
+	endBlock := finalizedBlock.Uint64()
+
+	// Search for `AssetsUnlockConfirmed` events for the given unlock sequences.
+	// Limit the search to finalized blocks.
+	confirmedEvents, err := s.bridgeContract.PastAssetsUnlockConfirmedEvents(
+		0, // TODO: Would setting startBlock to a lower value improve performance?
```

Comment:
When searching for recent `AssetsUnlocked` events on Mezo we look back `~1 month`.
We could do the same here - limit the search for `AssetsUnlockConfirmed` events on Ethereum to `~1 month`, since `AssetsUnlockConfirmed` for the given entry can only happen after such events is created on Mezo.

The problem, however, is we need to have a block number, rather than a timestamp. We can estimate which block on Ethereum was 1 month ago, but we could in theory miss some `AssetsUnlockConfirmed` events if the blocks on Ethereum were particularly short recently. Therefore, we would need some margin, e.g. look `1.5 months` back.

Reply:
A wide block span is always problematic for Ethereum RPC as providers typically impose 10k block limits and we need to use partial fetch. However, in this particular case, shouldn't we rely only on the sequences filter? Let's also split that to batches just in case and fetch per e.g. 1000 sequences (many providers limit responses to 10k logs). Then, in the next steps, we consider all events with block LTE finalizedBlock as confirmed and those with block GT finalizedBlock as unconfirmed. WDYT?